mono/metadata/icall.c

```diff
@@ -289,19 +293,23 @@ ves_icall_System_Array_SetValueImpl (MonoArray *arr, MonoObject *value, guint32
 	}
 
 	if (!ec->valuetype) {
-		if (!mono_object_isinst (value, ec))
+		if (!mono_object_isinst_checked (value, ec, &error))
 			INVALID_CAST;
 		mono_gc_wbarrier_set_arrayref (arr, ea, (MonoObject*)value);
 		return;
 	}
 
-	if (mono_object_isinst (value, ec)) {
+	if (mono_object_isinst_checked (value, ec, &error)) {
 		if (ec->has_references)
 			mono_value_copy (ea, (char*)value + sizeof (MonoObject), ec);
 		else
 			mono_gc_memmove_atomic (ea, (char *)value + sizeof (MonoObject), esize);
 		return;
 	}
+	if (!is_ok (&error)) {
+		mono_error_cleanup (&error);
```

Comment:
This is cleaning up the same error twice.

We should kill that INVALID_CAST macro and replace with something that takes no implicit state.